### PR TITLE
Handle np.array type in the case of unspecified dtype

### LINKED ISSUE #1152: Comparison with None fails with AttributeError
**Describe the bug**
Comparison between `None` or `Nil` and another object causes an `AttributeError`.

**To Reproduce**
Provide code to reproduce the behavior:
```python3
if __name__ == "__main__":
    print(None == 1)
```

**Error message**
```bash/fortran/c
Traceback (most recent call last):
  ...
  File "/home/pyccel/pyccel/parser/semantic.py", line 1922, in _visit_PyccelOperator
    return self._create_PyccelOperator(expr, args)
  File "/home/pyccel/pyccel/parser/semantic.py", line 648, in _create_PyccelOperator
    expr_new = type(expr)(*visited_args)
  File "/home/pyccel/pyccel/ast/operators.py", line 364, in __init__
    super().__init__(arg1, arg2)
  File "/home/pyccel/pyccel/ast/operators.py", line 128, in __init__
    self._set_shape_rank()
  File "/home/pyccel/pyccel/ast/operators.py", line 137, in _set_shape_rank
    self._shape, self._rank = self._calculate_shape_rank(*self._args)  # pylint: disable=no-member
  File "/home/pyccel/pyccel/ast/operators.py", line 437, in _calculate_shape_rank
    strs = [a for a in args if a.dtype is NativeString()]
  File "/home/pyccel/pyccel/ast/operators.py", line 437, in <listcomp>
    strs = [a for a in args if a.dtype is NativeString()]
AttributeError: 'Nil' object has no attribute 'dtype'
```

**Expected behavior**
It should print:
```bash/fortran/c
False
```

### PR DESCRIPTION
**Pull request by Pinkyboi**
Fix #1152: `np.array` now chooses the correct `dtype` from the type of its arguments:

- Extract `dtype` from `fill_value` without calling `process_dtype` in `NumpyFull` and `NumpyArray`.
- Change `NumpyZerosLike`, `NumpyOnesLike`, `NumpyEmptyLike`, `NumpyFullLike` to get `dtype` from passed argument if dtype not specified.
- Change `NumpyEmpty` to handle `NativeNumeric` dtype.

### CODE REVIEW

File: pyccel/ast/numpyext.py [hunk 1]
```diff
@@ -367,7 +367,7 @@ def __init__(self, arg, dtype=None, order='C'):
 
         # Verify dtype and get precision
         if dtype is None:
-            dtype = arg.dtype
+            dtype = DtypePrecisionToCastFunction[str(arg.dtype).capitalize()][arg.precision]
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:370]: In other places where this dictionary is accessed the key is: `arg.dtype.name` are you sure this will always give the same result as `str(arg.dtype).capitalize()`?

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:370]: I didn't notice that attribute, changing `str(arg.dtype).capitalize()` for `arg.dtype.name` seems like the best thing to do.

**Review by aihya — APPROVED** [review]

**Comment by Pinkyboi** [comment]
@EmilyBourne would it be a good idea to add a type check in `process_dtype` in case of unrecognized types ?
Something like the following:
```python
    if dtype not in dtype_registry:
        raise TypeError('Unknown type of  %s.' % dtype)
```

**Comment by EmilyBourne** [comment]
> @EmilyBourne would it be a good idea to add a type check in `process_dtype` in case of unrecognized types ? Something like the following:
> 
> ```python
>     if dtype not in dtype_registry:
>         raise TypeError('Unknown type of  %s.' % dtype)
> ```

I think that could be useful. I would also keep this commit: https://github.com/pyccel/pyccel/pull/1166/commits/cb23ee541ae54d32867e341eb8df49ab44e20a34

**Comment by Pinkyboi** [comment]
> I think that could be useful. I would also keep this commit: https://github.com/pyccel/pyccel/commit/cb23ee541ae54d32867e341eb8df49ab44e20a34

This commit caused some issues with pyccel tests that's why i discarded it.

**Comment by EmilyBourne** [comment]
> > I think that could be useful. I would also keep this commit: https://github.com/pyccel/pyccel/commit/cb23ee541ae54d32867e341eb8df49ab44e20a34
> 
> This commit caused some issues with pyccel tests that's why i discarded it.

That is because there are several other classes which use the `process_dtype` function but which aren't clear about the requested precision. Ideally I would rather you fixed the default argument in these classes

**Review by azouiten — APPROVED** [review]

**Comment by Pinkyboi** [comment]
@EmilyBourne I change `NumpyAutoFill` for it `dtype` to be recognized in `process_dtype` by making the default `dtype` a `NativeFloat` instead of an `str`.
I'm not fully satisfied by this fix, another way to do it is just to add str handling in `process_dtype` with something like:
```python
elif type(dtype) in [str, LiteralString]:
  dtype = str(dtype).replace('\'', '').lower()
```
Which one do you think is more suited ?

**Comment by EmilyBourne** [comment]
> @EmilyBourne I change `NumpyAutoFill` for it `dtype` to be recognized in `process_dtype` by making the default `dtype` a `NativeFloat` instead of an `str`.
> I'm not fully satisfied by this fix, another way to do it is just to add str handling in `process_dtype` with something like:
> ```python
> elif type(dtype) in [str, LiteralString]:
>   dtype = str(dtype).replace('\'', '').lower()
> ```
> Which one do you think is more suited ?

I prefer the solution with `PythonFloat` but doing `elif isinstance(dtype, (str, LiteralString)):` sounds useful in any case

**Comment by EmilyBourne** [comment]
> @EmilyBourne I change `NumpyAutoFill` for it `dtype` to be recognized in `process_dtype` by making the default `dtype` a `NativeFloat` instead of an `str`.
> I'm not fully satisfied by this fix, another way to do it is just to add str handling in `process_dtype` with something like:
> ```python
> elif type(dtype) in [str, LiteralString]:
>   dtype = str(dtype).replace('\'', '').lower()
> ```
> Which one do you think is more suited ?

I don't have a strong preference, but I would use `PythonFloat` rather than `NativeFloat`.
I think that adding `elif isinstance(dtype, (str, LiteralString))` could be useful even if you pick the first option

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
Please add at least 1 test that was failing before this PR

**Comment by Pinkyboi** [comment]
@EmilyBourne  I added added a test  in [58e2d8b](https://github.com/pyccel/pyccel/pull/1166/commits/58e2d8bada983aabcaf8c306dc1eb77ffb1944ee).

**Comment by Pinkyboi** [comment]
@EmilyBourne the test was firstly added in [58e2d8b](https://github.com/pyccel/pyccel/pull/1166/commits/58e2d8bada983aabcaf8c306dc1eb77ffb1944ee).

File: pyccel/ast/numpyext.py [hunk 2]
```diff
@@ -890,9 +894,10 @@ def __init__(self, shape, fill_value, dtype=None, order='C'):
         # If there is no dtype, extract it from fill_value
         # TODO: must get dtype from an annotated node
         if not dtype:
-            dtype = fill_value.dtype
-
+            dtype = DtypePrecisionToCastFunction[fill_value.dtype.name][fill_value.precision]
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:897]: Does `process_dtype` need to be called if the type is deduced from the array elements?

File: pyccel/ast/numpyext.py [hunk 3]
```diff
@@ -890,9 +894,10 @@ def __init__(self, shape, fill_value, dtype=None, order='C'):
...
         # Verify dtype and get precision
+        if dtype in NativeNumeric:
+            dtype = DtypePrecisionToCastFunction[dtype.name][-1]
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:900]: Why is this needed?

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:900]: In the way i implemented it it's needed because `NumpyFill` uses a `NativeNumeric` as a default value, this could be avoided if i reverted the default value back to an `str`.

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:900]: I think that would be more suited actually.

File: pyccel/ast/numpyext.py (hunk 2 above)
Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:897]: Actually not.

File: pyccel/ast/numpyext.py (hunk 3 above)
Comment (EmilyBourne) [reply pyccel/ast/numpyext.py]: `PythonFloat` is not a `NativeNumeric`

Comment (EmilyBourne) [reply pyccel/ast/numpyext.py]: Reverting the default value is also fine though. I think the original default is clearer

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py]: Oh yeah, that one's from the original change i forgot to remove.

File: pyccel/ast/numpyext.py (hunk 2 above)
Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:897]: In that case I would rather we set the value in an if/else block. It is unnecessarily costly to use the desired dtype and precision to find the cast function which will ensure that `process_dtype` returns the desired dtype and precision

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:897]: When actually coding that if else statement, i've found myself repeating a lot of what's done in `process_dtype` especially since  a different treatment of `dtype` is done depending on the precision (python types and numpy types) , so maybe dtype and precision calculation should still pass by `process_type` ?

File: pyccel/ast/numpyext.py (hunk 3 above)
Comment (Pinkyboi) [reply pyccel/ast/numpyext.py]: When i checked that condition again, i remembered that i left it because it solved the issue of classes like `NumpyImag`,
passing a `NativeNumeric` type as a specified dtype.

https://github.com/pyccel/pyccel/blob/a88889df050aed67343c06cc38532612b017ecca/pyccel/ast/numpyext.py#L216-L231

File: pyccel/ast/numpyext.py (hunk 2 above)
Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:897]: Maybe you misunderstood me then. I was thinking of an if/else like:
```
if not dtype:
    dtype = fill_value.dtype
    precision = fill_value.precision
else:
    dtype, precision = process_type(dtype)
```

File: pyccel/ast/numpyext.py (hunk 3 above)
Comment (EmilyBourne) [reply pyccel/ast/numpyext.py]: Ah ok. That makes sense. Does `NumpyImag` set the correct precision though? Presumably this handling means it also uses default precision even if we are taking the imaginary part of a `complex64`?

We need to be sure if there is an error because we are missing the handling of `NumericNative` or because the function triggering the error isn't being clear enough about the requested precision

File: pyccel/ast/numpyext.py (hunk 2 above)
Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:897]: Oh yeah that works ! assuming that the precision is valid.
```python
        if not dtype:
            dtype = fill_value.dtype
            precision = fill_value.precision
            if precision == -1:
                precision = default_precision[str(dtype)]
        else:
            dtype, precision = process_dtype(dtype)
```
Something like that maybe ?

Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:897]: Or maybe:
```python
if not dtype:
    dtype = fill_value.dtype
    precision = get_final_precision(fill_value.precision)
else:
    dtype, precision = process_type(dtype)
```

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:897]: Yep !

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py]: Fixed in [acd42c3](https://github.com/pyccel/pyccel/pull/1166/commits/acd42c3d3e928a51a9188f6b868e3068f71b83ca).

File: pyccel/ast/numpyext.py (hunk 3 above)
Comment (Pinkyboi) [reply pyccel/ast/numpyext.py]: Fixed in [4d429e8](https://github.com/pyccel/pyccel/pull/1166/commits/4d429e8cee532f46d505f09b3c295b7852c9b27b).

File: pyccel/ast/numpyext.py [hunk 4]
```diff
@@ -366,9 +372,12 @@ def __init__(self, arg, dtype=None, order='C'):
             raise TypeError('we only accept homogeneous arguments')
 
         # Verify dtype and get precision
-        if dtype is None:
+        if not dtype:
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:375]: Why did you change this test? Are you sure that `dtype` can't have a value which could evaluate to `False` ?

**Review by EmilyBourne — CHANGES_REQUESTED** [review]
You seem to have fixed a lot more bugs than you first planned to. Could you add some tests which check the dtype for these cases please

File: pyccel/ast/numpyext.py (hunk 4 above)
Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:375]: Can it have a value that can be evaluated as `False` ? i was unaware of that, that would apply to all dtype checks i added.

Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:375]: `not dtype` is equivalent to `not bool(dtype)`. Some things return `False` when you do `bool(dtype)` e.g:
- empty lists
- `LiteralFalse`
- empty string?

So I just want to check if this handling is intentional or if a comparison with `None` would be more accurate

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:375]: Now that you mention it, i can see that i would more accurate to use a comparison with `None` and throw an error for the case you mentioned (done by `process_dtype`).

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py]: Fixed in [bdb80e6](https://github.com/pyccel/pyccel/pull/1166/commits/bdb80e6707ae29560e8ae000d3ceb5394f4e3017).

**Comment by Pinkyboi** [comment]
@EmilyBourne i added some additional tests in [6ac1b1c](https://github.com/pyccel/pyccel/pull/1166/commits/6ac1b1ce0b6f810658dce8e22ec12048660cf90d) but still not really sure why the Windows test fails.

**Comment by EmilyBourne** [comment]
> @EmilyBourne i added some additional tests in [6ac1b1c](https://github.com/pyccel/pyccel/pull/1166/commits/6ac1b1ce0b6f810658dce8e22ec12048660cf90d) but still not really sure why the Windows test fails.

I don't see a failing test. Occasionally tests fail because the random numbers generated give a badly conditioned problem

**Comment by EmilyBourne** [comment]
> @EmilyBourne i added some additional tests in [6ac1b1c](https://github.com/pyccel/pyccel/pull/1166/commits/6ac1b1ce0b6f810658dce8e22ec12048660cf90d) but still not really sure why the Windows test fails.

I don't see a failing test. Occasionally tests fail because the random numbers generated give a badly conditioned problem

**Comment by Pinkyboi** [comment]
@EmilyBourne i did run the Windows test again and it solved the issue.

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 5]
```diff
@@ -3127,84 +3451,150 @@ def create_zeros_like_shape_F(n,m):
...
+
     f_real_complex64   = epyccel(create_zeros_like_val_complex64, language = language)
-    assert(isclose(     f_real_complex64()  ,      create_zeros_like_val_complex64(), rtol=RTOL, atol=ATOL))
     assert matching_types(f_real_complex64(), create_zeros_like_val_complex64())
 
+    f_complex128_auto   = epyccel(create_zeros_like_val_complex128_auto, language = language)
+    assert matching_types(f_complex128_auto(), create_zeros_like_val_complex128_auto())
+
     f_real_complex128   = epyccel(create_zeros_like_val_complex128, language = language)
-    assert(isclose(     f_real_complex128() ,      create_zeros_like_val_complex128(), rtol=RTOL, atol=ATOL))
```
Comment (EmilyBourne) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:3207]: Why did you decide to remove the value tests?

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 5 above)
Comment (Pinkyboi) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py:3207]: that was a mistake, ill add it back.

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 6]
```diff
@@ -1413,6 +1413,90 @@ def create_full_val_real_complex128(val):
...
+    @types('int')
+    def create_full_val_int(val):
+        from numpy import full
+        a = full(3,val)
+        return a[0]
+    @types('float')
+    def create_full_val_float(val):
+        from numpy import full
+        a = full(3,val)
+        return a[0]
```
[the remark this answers is not in the document]
Comment (Pinkyboi) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py:1426]: I had troubles working with multiple type decorators with `python` as `language`, ill try to investigate on the cause of that but those multiple function worked like a workaround.

File: pyccel/ast/numpyext.py [hunk 7]
```diff
@@ -934,6 +943,12 @@ class NumpyEmpty(NumpyAutoFill):
     """
     __slots__ = ()
     name = 'empty'
+
+    def __init__(self, shape, dtype='float', order='C'):
+        if dtype in NativeNumeric:
+            precision = default_precision[str(dtype)]
```
[the remark this answers is not in the document]
Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:949]: `default_precision[str_dtype(dtype)]` seem better suited ill change it.

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py:1426]: This is normal. There are several test which have `xfail` or `skip` tags for this case, and there is an issue (#885 ) to fix them one day. Translating to python doesn't accelerate anything, it just acts as a guide to see the internal changes pyccel may make, so skipping these tests is not a huge problem

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 5 above)
Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py:3207]: It primarily checks the type, but extra tests are never a problem ;)

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 8]
```diff
@@ -1413,6 +1413,72 @@ def create_full_val_real_complex128(val):
     assert(isclose(     f_real_complex128(val_float)       ,      create_full_val_real_complex128(val_float), rtol=RTOL, atol=ATOL))
     assert matching_types(f_real_complex128(val_float), create_full_val_real_complex128(val_float))
 
+@pytest.mark.parametrize( 'language', (
+        pytest.param("fortran", marks = pytest.mark.fortran),
+        pytest.param("c", marks = pytest.mark.c),
+        pytest.param("python", marks = [
+            pytest.mark.skip,
```
Comment (EmilyBourne) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:1420]: Please add a reason to your skip mark. See [here](https://github.com/pyccel/pyccel/blob/master/tests/epyccel/recognised_functions/test_numpy_funcs.py#L3528-L3537) for an example

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 9]
```diff
@@ -1413,6 +1413,72 @@ def create_full_val_real_complex128(val):
...
+    @types('float64')
+    @types('complex64')
+    @types('complex128')
+    def create_full_val_auto(val):
+        from numpy import full
+        a = full(3,val)
+        return a[0]
+
+    integer   = randint(low = iinfo('int').min,   high = iinfo('int').max,   dtype=int)
+    integer32 = randint(low = iinfo('int32').min, high = iinfo('int32').max, dtype=np.int32)
```
Comment (EmilyBourne) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:1441]: Values like `iinfo('int').min` are already defined at the top of this file. See [here](https://github.com/pyccel/pyccel/blob/master/tests/epyccel/recognised_functions/test_numpy_funcs.py#L11-L33). Please use them rather than recalculating them

**Review by EmilyBourne — APPROVED** [review]

**Comment by yguclu** [comment]
@Pinkyboi Good job! Can you please extend the PR description with a list of all the relevant modifications?

File: tests/epyccel/modules/arrays.py [hunk 10]
```diff
@@ -7,6 +7,22 @@
...
+@types('int8', 'int8')
+@types('int16', 'int16')
+@types('int32', 'int32')
+@types('int64', 'int64')
+@types('float', 'float')
+@types('float32', 'float32')
+@types('float64', 'float64')
+@types('complex64', 'complex64')
+@types('complex128', 'complex128')
+def array_return_first_element(a, b):
```
Comment (yguclu) [inline tests/epyccel/modules/arrays.py:21]: Can't we handle all these cases using a template? See [here](https://github.com/pyccel/pyccel/blob/master/tutorial/templates.md#templates-using-decorators)

File: tests/epyccel/recognised_functions/test_numpy_funcs.py [hunk 11]
```diff
@@ -1413,6 +1413,73 @@ def create_full_val_real_complex128(val):
...
+def test_full_dtype_auto(language):
+    @types('int')
+    @types('float')
+    @types('complex')
+    @types('int32')
+    @types('float32')
+    @types('float64')
+    @types('complex64')
+    @types('complex128')
+    def create_full_val_auto(val):
```
Comment (yguclu) [inline tests/epyccel/recognised_functions/test_numpy_funcs.py:1436]: Same comment about templates

File: pyccel/ast/numpyext.py [hunk 12]
```diff
@@ -340,8 +342,12 @@ def process_dtype(dtype):
 				  NumpyComplex128, NumpyComplex64, NumpyFloat64, NumpyFloat32):
         # remove numpy prefix from dtype.name len("numpy") = 5
         dtype = dtype.__name__.lower()[5:]
+    elif isinstance(dtype, (LiteralString, str)):
+        dtype = str(dtype).replace('\'', '').lower()
+        if dtype not in dtype_registry:
+            raise TypeError(f'Unknown type of {dtype}.')
     else:
-        dtype            = str(dtype).replace('\'', '').lower()
+        raise TypeError(f'Unknown type of {dtype}.')
```
Comment (yguclu) [inline pyccel/ast/numpyext.py:350]: I have just noticed that this function (`process_dtype`) does not have a docstring. Would you mind writing one?

**Comment by Pinkyboi** [comment]
@yguclu I added what you requested, im not really confident about the new description or the docstring if you have any remarks please tell me about em.

File: tests/epyccel/modules/arrays.py (hunk 10 above)
Comment (EmilyBourne) [reply tests/epyccel/modules/arrays.py]: Fixed [5421f09](https://github.com/pyccel/pyccel/pull/1166/commits/5421f09d6e770384f5e0a2fad5dc8d9b72b025e5)

File: tests/epyccel/recognised_functions/test_numpy_funcs.py (hunk 11 above)
Comment (EmilyBourne) [reply tests/epyccel/recognised_functions/test_numpy_funcs.py]: Fixed [5421f09](https://github.com/pyccel/pyccel/pull/1166/commits/5421f09d6e770384f5e0a2fad5dc8d9b72b025e5)

File: pyccel/ast/numpyext.py [hunk 13]
```diff
@@ -328,6 +330,21 @@ class NumpyComplex128(NumpyComplex):
...
+        dtype     : Datatype
+
+    Raises:
+        TypeError: In the case of unrecognized argument type.
+        TypeError: In the case of passed string argument not recognized as valid dtype.
+
+    Returns:
+        dtype     : Datatype
+        precision : int
+    """
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:346]: We try to use the same formatting for all our docstrings which allows them to be recognised by doxygen if we ever have enough documentation to make it worth adding automated documentation. See [here](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/core.py#L164-L188) for an example of a docstring.

After reading the docstring we should understand:
- [x] what the purpose of the function is
- [ ] In what context the function should be used
- [ ] What the expected type of each of the arguments are
- [ ] What each of the arguments is for
- [x] What the expected type of each of the results are

Currently we haven't added a section for possible errors but I think this could be useful so I would keep this.

I think the information missing from your docstring is:
- Argument type (If `dtype` is a `DataType` then a `TypeError` is raised so the type that you have written is not the expected type)
- Argument role
- The description could be expanded to explain where this function should be used (i.e. to handle `dtype` arguments of array creation functions). The explanation should be clear enough to avoid the confusion that you had when you started tackling this issue.

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:346]:
```
    """
    This function will take a dtype passed to a numpy array creation function
    and depending on the type of the passed argument, it will do the necessary treatments
    to extract the corresponding type and precision from it then return the result.

    This function could be useful when working with numpy creation function having a dtype argument,
    like numpy.array, nump.arrange, numpy.linespace...
    
    Parameters:
    ----------
        dtype     : PythonType | PyccelFunctionDef | String | Numpy Datatype | Python Builtin Datatype
        The actual dtype passed to the numpy function

    Raises:
    ----------
        TypeError : In the case of unrecognized argument type.
        TypeError : In the case of passed string argument not recognized as valid dtype.

    Returns:
    ----------
        dtype     : Datatype
        The Datatype corresponding to the passed dtype.
        precision : int
        The precision corresponding to the passed dtype.
    """
```
Is something like this better ?

Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:346]: That looks a lot better :)

Typo `linespace` -> `linspace`

What `PyccelFunctionDef` other than a `Numpy Datatype` or a `Python Builtin Datatype` can be used as an argument?

Comment (Pinkyboi) [reply pyccel/ast/numpyext.py:346]: From what i know it can only be one of those two.

Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:346]: Then I don't think you need to mention `PyccelFunctionDef` separately.
You should have one of:
```
dtype     : PythonType | PyccelFunctionDef | String
```
```
dtype     : PythonType | String | Numpy Datatype | Python Builtin Datatype
```
```
dtype     : PythonType | PyccelFunctionDef (Numpy Datatype or Python Builtin Datatype) | String
```

**Review by EmilyBourne — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Comment by yguclu** [comment]
Fix #1152: `np.array` now chooses the correct `dtype` from the type of its arguments:

- Extract `dtype` from `fill_value` without calling `process_dtype` in `NumpyFull` and `NumpyArray`.
- Change `NumpyZerosLike`, `NumpyOnesLike`, `NumpyEmptyLike`, `NumpyFullLike` to get `dtype` from passed argument if dtype not specified.
- Change `NumpyEmpty` to handle `NativeNumeric` dtype.
